vortex-file/src/generic.rs

```diff
@@ -348,6 +355,48 @@ where
     ret
 }
 
+struct CoalescingMetrics {
+    bytes_uncoalesced: Arc<Counter>,
+    bytes_coalesced: Arc<Counter>,
+    request_count_uncoalesced: Arc<Counter>,
+    request_count_coalesced: Arc<Counter>,
+}
+
+impl From<Arc<VortexMetrics>> for CoalescingMetrics {
+    fn from(metrics: Arc<VortexMetrics>) -> Self {
+        let byte_ranges = MetricId::new("vortex.scan.requests.bytes");
+        let requests = MetricId::new("vortex.scan.requests.count");
+        Self {
+            bytes_uncoalesced: metrics.counter(byte_ranges.clone().with_tag("kind", "uncoalesced")),
+            bytes_coalesced: metrics.counter(byte_ranges.with_tag("kind", "coalesced")),
+            request_count_uncoalesced: metrics
+                .counter(requests.clone().with_tag("kind", "uncoalesced")),
+            request_count_coalesced: metrics.counter(requests.with_tag("kind", "coalesced")),
+        }
+    }
+}
+
+impl CoalescingMetrics {
+    fn record(&self, req: &CoalescedSegmentRequest) {
+        // record request counts
+        self.request_count_coalesced.inc();
+        let signed_len = req.requests.len().try_into();
+        let _ = signed_len.map(|len| self.request_count_uncoalesced.add(len));
+
+        // record uncoalesced total byte requests vs coalesced
+        let signed_bytes = (req.byte_range.end - req.byte_range.start).try_into();
+        let _ = signed_bytes.map(|bytes| self.bytes_coalesced.add(bytes));
+        let _ = req
+            .requests
+            .iter()
+            .map(|req| req.range())
+            .map(|r| r.end - r.start)
+            .sum::<u64>()
+            .try_into()
+            .map(|bytes| self.bytes_uncoalesced.add(bytes));
```

Comment:
you can have a couple of `if let` here or do that Scala thing and treat the option as an iterator